[gree] `turbo`, `light`, `health`, `xfan` switches for supported models

nagyrobi opened the pull request:
# What does this implement/fix?

Some GREE units support a couple of additional features which are useful to be controlled through ESPHome (quotes from user manuals):

- To turn off indoor unit display night for complete room darkness, press Light button.
- Turbo button for maximum fan speed and fastest results.
- X-Fan button before turning off the AC (in cooling or dry mode), to prevent excess moisture that cause mold, mildew, and unpleasant odors. Indoor Fan will keep running for short period after you turn off the AC, to dry the blades.
- Health removes dust and germs from your room by ionizing the air flowing through the blades.

Term button refers to the button on the IR remote, but it's internally a switch, also reflected by the remote's own display. Press of the button toggles the feature ON and OFF in the unit - a switch in ESPHome is the perfect use case for this.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Developer breaking change (an API change that could break external components)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes https://github.com/esphome/feature-requests/issues/2762
- implementation based on Jesse's suggestion in https://discordapp.com/channels/429907082951524364/1254011151721955418/1254022129796775986

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#5698

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

Tested with 3 separate Gree AC units, works flawlessly on Geklink IR blasters.

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
climate:
  - platform: gree
    model: yan
    id: ir_klima
    name: My room AC
    visual:
      min_temperature: 18
      max_temperature: 30
      temperature_step:
        target_temperature: 1.0

switch:
  - platform: gree
    gree_id: ir_klima
    turbo:
      name: AC Turbo
    light:
      name: AC Lights
    health:
      name: AC Health
    xfan:
      name: AC X-Fan
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nagyrobi committed:
Default Gree switches to restore off

nagyrobi committed:
Use dedicated gree_id in Gree switch schema

nagyrobi committed:
Merge pull request #3 from nagyrobi/codex/add-switches-for-gree-climate-models

Set default icons and restore mode for Gree switches

nagyrobi committed:
Revert "Set default icons and restore mode for Gree switches"

nagyrobi committed:
Merge pull request #5 from nagyrobi/revert-3-codex/add-switches-for-gree-climate-models

Revert "Set default icons and restore mode for Gree switches"

nagyrobi committed:
Merge pull request #4 from nagyrobi/codex/add-switches-for-gree-climate-models-a52hnk

Use dedicated gree_id in Gree switch schema

nagyrobi committed:
Fix health switch constant import

nagyrobi committed:
Revert "Use dedicated gree_id in Gree switch schema"

nagyrobi committed:
Merge pull request #7 from nagyrobi/revert-4-codex/add-switches-for-gree-climate-models-a52hnk

Revert "Use dedicated gree_id in Gree switch schema"

nagyrobi committed:
Export gree namespace for switches

nagyrobi committed:
Handle mock models in Gree switch validation

nagyrobi committed:
Remove gree dependency requirement for switches

nagyrobi committed:
Validate model compatibility for Gree switches

nagyrobi committed:
Normalize Gree switch model validation

nagyrobi committed:
Adjust Gree health switch icon

nagyrobi committed:
Merge pull request #6 from nagyrobi/codex/add-switches-for-gree-climate-models-kegqix

Fix health switch constant import

nagyrobi committed:
Restore Gree switch state on boot

nagyrobi committed:
Revert "Fix health switch constant import"

nagyrobi committed:
Merge pull request #9 from nagyrobi/revert-6-codex/add-switches-for-gree-climate-models-kegqix

Revert "Fix health switch constant import"

nagyrobi committed:
Merge pull request #8 from nagyrobi/codex/add-switches-for-gree-climate-models-rpmsxh

Restore Gree switch state on boot

nagyrobi committed:
Revert "Restore Gree switch state on boot"

nagyrobi committed:
Merge pull request #10 from nagyrobi/revert-8-codex/add-switches-for-gree-climate-models-rpmsxh

Revert "Restore Gree switch state on boot"

nagyrobi committed:
Define log tag for Gree switches

nagyrobi committed:
Merge pull request #11 from nagyrobi/codex/add-switches-for-gree-climate-models-wjhiky

Define log tag for Gree switches

nagyrobi committed:
Revert "Define log tag for Gree switches"

nagyrobi committed:
Merge pull request #12 from nagyrobi/revert-11-codex/add-switches-for-gree-climate-models-wjhiky

Revert "Define log tag for Gree switches"

nagyrobi committed:
Publish Gree switch states when set

nagyrobi committed:
Avoid overriding restored Gree switch states

nagyrobi committed:
Clarify switch bit mirroring for Gree models

nagyrobi committed:
Register Gree switches as components

nagyrobi committed:
Make Gree switches components for registration

nagyrobi committed:
Merge pull request #13 from nagyrobi/codex/add-switches-for-gree-climate-models-ifmoca

Publish Gree switch states when set

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

nagyrobi committed:
Fix gree switch constant definitions

nagyrobi committed:
Merge pull request #14 from nagyrobi/codex/fix-constant-naming-and-imports-in-gree-switch

Fix gree switch constant definitions

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

nagyrobi committed:
Update CODEOWNERS

nagyrobi committed:
Change CODEOWNERS for gree switch component

nagyrobi committed:
Merge pull request #15 from nagyrobi/codex/update-codeowners-file

Update CODEOWNERS

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

nagyrobi committed:
Update CODEOWNERS

nagyrobi committed:
Merge pull request #16 from nagyrobi/codex/update-codeowners-file-fkmcjw

Update CODEOWNERS file

nagyrobi committed:
Gree switch tests

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into gree-climate-with-switches

clydebarrow commented on esphome/components/gree/gree.cpp:
```diff
@@ -52,6 +123,34 @@ void GreeClimate::transmit_state() {
+    if (this->turbo_mode_) {
+      remote_state[2] |= (1 << 4);  // Set bit 4 (TURBO ON)
+    } else {
+      remote_state[2] &= ~(1 << 4);  // Clear bit 4 (TURBO OFF)
```
Shouldn't be needed - bit 4 is never set before this point.

clydebarrow committed:
Combine mode bits into single member;
Remove redundant publish_state in parent component, and remove redundant
switch references.

nagyrobi committed:
Merge branch 'dev' into gree-climate-with-switches

clydebarrow committed:
Abstract the 4 switch types into one.

clydebarrow committed:
Merge branch 'gree-climate-with-switches' of https://github.com/nagyrobi/esphome into gree-climate-with-switches

clydebarrow committed:
codeowners

clydebarrow committed:
Improve error message

clydebarrow committed:
Add entity properties

clydebarrow committed:
Add Device properties
Fix case-sensitivity for model

clydebarrow committed:
Merge branch 'dev' into gree-climate-with-switches

clydebarrow committed:
Cleanup
